### ISSUE TITLE
Could I fuse two pretrained models in one to train a model that has defferent struct from pretrained model?

### ISSUE THREAD

**Issue by jajajajaja121** [open-mmlab/mmdetection#3287]
For example, I use FCOS head and Retina head into one model to predict bouding box, and I want to load whole fcos pretrainmodels, and the parameters of Retina Head  into my model, that will work? At present, I have successfully load the pretrained prameters into my model, but the result is terrible, I don't know wether my code has some where wrong, or load parameters in this way is incorrect.

**Comment by jajajajaja121** [comment]
My model has a low cost, but a terrible output, when I watch the predict result, I can see, most of the predict result is toothbrush... by the way, I train my model with 5000 images from COCO dataset, and toothbrush is the last label of the whole class.

**Comment by jajajajaja121** [comment]
`@force_fp32(apply_to=('cls_scores', 'bbox_preds', 'centernesses'))
    def get_bboxes(self,
                   cls_scores,
                   bbox_preds,
                   centernesses,
                   img_metas,
                   cfg=None,
                   rescale=None):
        assert len(cls_scores) == len(bbox_preds)
        device = cls_scores[0][0].device
        num_levels = len(cls_scores)
        cls_scores_AB = [cls_score[0] for cls_score in cls_scores]#得到anchor base的分类得分
        cls_scores_AF = [cls_score[1] for cls_score in cls_scores]#得到anchor free的分类得分
        bbox_preds_AB = [bbox_pred[0] for bbox_pred in bbox_preds]#得到anchor base的回归得分
        bbox_preds_AF = [bbox_pred[1] for bbox_pred in bbox_preds]#得到anchor free的回归得分
        featmap_sizes = [featmap.size()[-2:] for featmap in cls_scores_AF]#得到特征图的尺寸
        mlvl_points = self.get_points(featmap_sizes, bbox_preds_AF[0].dtype,
                                      bbox_preds_AF[0].device)
        result_list = []
        featmap_sizes = [featmap.size()[-2:] for featmap in cls_scores_AF]#得到各级特征图的尺寸
        anchors = self.anchor_base.anchor_generator.grid_anchors(#计算anchor,返回值为各级特征上anchor的坐标
            featmap_sizes, device=device)
        for img_id in range(len(img_metas)):
            cls_score_list_AF = [
                cls_scores_AF[i][img_id].detach() for i in range(num_levels)#将一个batch图片的分类得分放到一起
            ]
            cls_score_list_AB = [
                cls_scores_AB[i][img_id].detach() for i in range(num_levels)
            ]
            cls_score_list=[cls_score_list_AF,cls_score_list_AB]#两种head的分类得分打包
            bbox_pred_list_AF = [
                bbox_preds_AF[i][img_id].detach() for i in range(num_levels)
            ]
            bbox_pred_list_AB = [
                bbox_preds_AB[i][img_id].detach() for i in range(num_levels)
            ]
            bbox_pred_list=[bbox_pred_list_AF,bbox_pred_list_AB]#两种head的回归得分打包
            centerness_pred_list_AF = [
                centernesses[i][img_id].detach() for i in range(num_levels)#将一个batch的centerness放到一起
            ]
            centerness_pred_list_AB=[
                centernesses[i][img_id].repeat(9,1,1).detach() for i in range(num_levels)
            ]
            centerness_pred_list=[centerness_pred_list_AF,centerness_pred_list_AB]#将两种head的centerness放到一起
            # centernesses_AB=[centerness for centerness in bbox_pred_list_AB]
            # for i,centerness in enumerate(centernesses_AB):
            #     centernesses_AB[i]=torch.ones_like(centerness)
            #     centernesses_AB[i]=centernesses_AB[i][0]
            #     centernesses_AB[i]=torch.unsqueeze(centernesses_AB[i],0)
            # centerness_pred_list=[centerness_pred_list,centernesses_AB]
            img_shape = img_metas[img_id]['img_shape']#图片尺寸
            scale_factor = img_metas[img_id]['scale_factor']
            det_bboxes,det_labels = self._get_bboxes_single(cls_score_list,#得到最终的bbox预测结果
                                                 bbox_pred_list,
                                                 centerness_pred_list,
                                                 mlvl_points, img_shape,
                                                 scale_factor, cfg, anchors,rescale)
            result_list.append([det_bboxes,det_labels])
        return result_list

    def _get_bboxes_single(self,
                           cls_scores,
                           bbox_preds,
                           centernesses,
                           mlvl_points,
                           img_shape,
                           scale_factor,
                           cfg,
                           anchors,
                           rescale=False):
        cfg = self.test_cfg if cfg is None else cfg
        assert len(cls_scores[0]) == len(bbox_preds[0]) == len(mlvl_points)
        mlvl_bboxes = []
        mlvl_scores = []
        mlvl_centerness = []
        for cls_score, bbox_pred, centerness, points in zip(
                cls_scores[0], bbox_preds[0], centernesses[0], mlvl_points):#这个for循环得到anchor free分支的预测结果
            assert cls_score.size()[-2:] == bbox_pred.size()[-2:]
            scores = cls_score.permute(1, 2, 0).reshape(
                -1, self.cls_out_channels).sigmoid()
            centerness = centerness.permute(1, 2, 0).reshape(-1).sigmoid()
            bbox_pred = bbox_pred.permute(1, 2, 0).reshape(-1, 4)
            nms_pre = cfg.get('nms_pre', -1)
            if nms_pre > 0 and scores.shape[0] > nms_pre:
                max_scores, _ = (scores * centerness[:, None]).max(dim=1)
                _, topk_inds = max_scores.topk(nms_pre)
                points = points[topk_inds, :]
                bbox_pred = bbox_pred[topk_inds, :]
                scores = scores[topk_inds, :]
                centerness = centerness[topk_inds]
            bboxes = distance2bbox(points, bbox_pred, max_shape=img_shape)
            mlvl_bboxes.append(bboxes)
            mlvl_scores.append(scores)
            mlvl_centerness.append(centerness)
        for cls_score_AB, bbox_pred_AB, points,anchor,centerness_AB in zip(#这个for循环得到anchor base分支的预测结果
                cls_scores[1], bbox_preds[1], mlvl_points,anchors,centernesses[1]):
            assert cls_score_AB.size()[-2:] == bbox_pred_AB.size()[-2:]
            scores = cls_score_AB.permute(1, 2, 0).reshape(
                -1, self.cls_out_channels).sigmoid()
            bbox_pred_AB = bbox_pred_AB.permute(1, 2, 0).reshape(-1, 4)
            centerness_AB = centerness_AB.permute(1, 2, 0).reshape(-1).sigmoid()#计算centerness
            nms_pre = cfg.get('nms_pre', -1)
            if nms_pre > 0 and scores.shape[0] > nms_pre:#当bbox太多时,则将得分最高的1000个取出
                max_scores, _ = (scores * 1).max(dim=1)
                _, topk_inds = max_scores.topk(nms_pre)
                bbox_pred_AB = bbox_pred_AB[topk_inds, :]
                scores = scores[topk_inds, :]
                anchor = anchor[topk_inds, :]
                centerness_AB=centerness_AB[topk_inds]
            bboxes = self.anchor_base.bbox_coder.decode(anchor, bbox_pred_AB)#将bbox映射为原图上的坐标
            mlvl_bboxes.append(bboxes)
            mlvl_scores.append(scores)
            mlvl_centerness.append(centerness_AB)
        mlvl_bboxes = torch.cat(mlvl_bboxes)#将所有特征层的bbox融合到同一个四维的张量中
        if rescale:
            mlvl_bboxes /= mlvl_bboxes.new_tensor(scale_factor)
        mlvl_scores = torch.cat(mlvl_scores)#所有特征层的得分融合到一个一维张量中
        padding = mlvl_scores.new_zeros(mlvl_scores.shape[0], 1)#给分类得分补零
        # remind that we set FG labels to [0, num_class-1] since mmdet v2.0
        # BG cat_id: num_class
        mlvl_scores = torch.cat([mlvl_scores, padding], dim=1)
        mlvl_centerness_all = torch.cat(mlvl_centerness)#将所有centerness融合到一个张量中
        # mlvl_centerness_all = torch.ones_like(mlvl_bboxes)[:, 0]
        # mlvl_centerness_all[:len(mlvl_centerness)] = mlvl_centerness
        # mlvl_centerness_all=(mlvl_centerness.repeat(1,10)).squeeze()
        det_bboxes, det_labels = multiclass_nms(#nms输出最终结果
            mlvl_bboxes,
            mlvl_scores,
            cfg.score_thr,
            cfg.nms,
            cfg.max_per_img,
            score_factors=mlvl_centerness_all)
        return det_bboxes, det_labels#返回最终结果`

**Comment by jajajajaja121** [comment]
Here is my code, that get bbox when testing models.

**Comment by hhaAndroid** [comment]
Do you want to do multi-model output fusion? If so, you can infernece about FCOS and RetinaNet separately, and then merge the results.
